src/Microsoft.AspNetCore.Routing/RouteValuesBasedEndpointFinder.cs

```diff
@@ -0,0 +1,170 @@
+using Microsoft.Extensions.ObjectPool;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal class RouteValuesBasedEndpointFinder : IEndpointFinder<RouteValuesContext>
+    {
+        private readonly CompositeEndpointDataSource _endpointDataSource;
+        private readonly IInlineConstraintResolver _inlineConstraintResolver;
+        private readonly ObjectPool<UriBuildingContext> _objectPool;
+        private readonly LinkGenerationDecisionTree _linkGenerationTree;
+
+        public RouteValuesBasedEndpointFinder(
+            CompositeEndpointDataSource endpointDataSource,
+            ObjectPool<UriBuildingContext> objectPool)
+        {
+            // TODO: register for change events?
+            _endpointDataSource = endpointDataSource;
+            _objectPool = objectPool;
+
+            // TODO: Review: IInlineConstraintResolver is registered as Transient, where as we want this finder to be
+            // registered as singleton.
+            _inlineConstraintResolver = new DefaultInlineConstraintResolver(Options.Create(new RouteOptions()));
+
+            var outboundRouteEntries = GetOutboundRouteEntries();
+
+            var outboundMatches = GetOutboundMatches(outboundRouteEntries);
+
+            _linkGenerationTree = new LinkGenerationDecisionTree(outboundMatches.ToArray());
+        }
+
+        public IEnumerable<Endpoint> FindEndpoints(RouteValuesContext context)
+        {
+            var matches = _linkGenerationTree.GetMatches(context.ExplicitValues, context.AmbientValues);
+
+            return matches
+                .Select(r => r.Match.Entry)
+                .OfType<OutboundRouteEntryWithEndpoint>()
+                .Select(r => r.Endpoint);
+        }
+
+        private IList<OutboundRouteEntryWithEndpoint> GetOutboundRouteEntries()
+        {
+            var outboundRouteEntries = new List<OutboundRouteEntryWithEndpoint>();
+
+            var endpoints = _endpointDataSource.Endpoints.OfType<MatcherEndpoint>();
+            foreach (var endpoint in endpoints)
+            {
+                var requiredValuesMetadata = endpoint.Metadata.GetMetadata<IRequiredLinkValuesMetadata>();
+                var requiredLinkValues = requiredValuesMetadata?.Values;
+                if (requiredLinkValues == null)
+                {
+                    requiredLinkValues = new RouteValueDictionary();
+                }
+
+                var entry = new OutboundRouteEntryWithEndpoint()
+                {
+                    Handler = NullRouter.Instance,
+                    Order = endpoint.Order,
+                    Precedence = RoutePrecedence.ComputeOutbound(endpoint.ParsedTemlate),
+                    RequiredLinkValues = requiredLinkValues,
+                    // TODO: handle routes with names. What does it mean to have a route name on 
+                    // a parameterized template like {controller}/{action}/{id?}, name="default" ?.
+                    RouteName = null,
```

Comment:
As discussed, we can remove the validation that prevents this from being possible.